File: packages/core/src/routing/graph.ts
import type { CircuitPack, ZoneState } from '@crowdflow/contracts';
import { ASSUMED_ROUTE_CACHE_ENTRIES, FREE_FLOW_SPEED_MS, isOpenDuring, isTrustworthy } from '@crowdflow/contracts';
import { MIN_SPEED_MS } from '../state/flow.js';

export const CONGESTION_WEIGHT = 2.5;
export const CRITICAL_WEIGHT = 12;
export const UNTRUSTED_WIDTH_PENALTY = 1.15;
export const AVOID_PENALTY = 25;
export const PREFER_DISCOUNT = 0.6;

export interface RouteResult {
  path: string[];
  edge_ids: string[];
  cost_s: number;
  distance_m: number;
  eta_s: number;
  rejected_reason?: string;
}

export interface OperationalRestrictions {
  closedEdges?: Iterable<string>;
  closedZones?: Iterable<string>;
  edgeCapacity?: Iterable<[string, number]>;
  zoneCapacity?: Iterable<[string, number]>;
}

type CacheKey = string;

export class VenueGraph {
  readonly pack: CircuitPack;
  sessionState: string | null;
  cacheHits = 0;
  cacheMisses = 0;
  private adjacency = new Map<string, Array<[string, string]>>();
  private closed = new Set<string>();
  private forbidden = new Set<string>();
  private forbiddenEdges = new Set<string>();
  private operationalClosedEdges = new Set<string>();
  private operationalClosedZones = new Set<string>();
  private edgeCapacityFactors = new Map<string, number>();
  private zoneCapacityFactors = new Map<string, number>();
  private cache = new Map<CacheKey, RouteResult>();
  private operationalRevision = 0;

  constructor(pack: CircuitPack, sessionState: string | null = null) {
    this.pack = pack;
    this.sessionState = sessionState;
    this.rebuild(sessionState);
  }

  rebuild(sessionState: string | null): void {
    this.sessionState = sessionState;
    this.cache.clear();
    this.closed.clear();
    for (const crossing of Object.values(this.pack.crossings ?? {})) {
      if (!isOpenDuring(crossing.availability ?? {}, sessionState)) this.closed.add(crossing.edge_id);
    }
    this.forbidden = new Set(this.pack.constraints?.never_route_through ?? []);
    this.forbiddenEdges = new Set(this.pack.constraints?.never_route_edges ?? []);
    this.adjacency = new Map(Object.keys(this.pack.zones ?? {}).map((id) => [id, []]));
    for (const [id, edge] of Object.entries(this.pack.edges ?? {})) {
      if (
        this.closed.has(id) ||
        this.forbiddenEdges.has(id) ||
        this.forbidden.has(edge.source) ||
        this.forbidden.has(edge.destination)
      )
        continue;
      this.adjacency.get(edge.source)?.push([edge.destination, id]);
      if (edge.bidirectional ?? true) this.adjacency.get(edge.destination)?.push([edge.source, id]);
    }
  }

  get routeCacheSize(): number { return this.cache.size; }
  get closedEdges(): Set<string> { return new Set(this.closed); }
  get forbiddenZones(): Set<string> { return new Set(this.forbidden); }
  get restrictionsRevision(): number { return this.operationalRevision; }

  setOperationalRestrictions(restrictions: OperationalRestrictions): void {
    this.operationalClosedEdges = new Set(restrictions.closedEdges ?? []);
    this.operationalClosedZones = new Set(restrictions.closedZones ?? []);
    this.edgeCapacityFactors = new Map(restrictions.edgeCapacity ?? []);
    this.zoneCapacityFactors = new Map(restrictions.zoneCapacity ?? []);
    for (const factor of [...this.edgeCapacityFactors.values(), ...this.zoneCapacityFactors.values()]) {
      if (!(factor > 0 && factor <= 1)) throw new Error('capacity factors must be in (0, 1]');
    }
    this.operationalRevision += 1;
    this.cache.clear();
  }

  operationalState(): { closed_edges: string[]; closed_zones: string[]; edge_capacity: Record<string, number>; zone_capacity: Record<string, number> } {
    return {
      closed_edges: [...this.operationalClosedEdges].sort(),
      closed_zones: [...this.operationalClosedZones].sort(),
      edge_capacity: Object.fromEntries([...this.edgeCapacityFactors].sort(([a], [b]) => a.localeCompare(b))),
      zone_capacity: Object.fromEntries([...this.zoneCapacityFactors].sort(([a], [b]) => a.localeCompare(b))),
    };
  }

  isEdgeAvailable(edgeId: string): boolean { return !this.closed.has(edgeId) && !this.operationalClosedEdges.has(edgeId); }
  isZoneAvailable(zoneId: string): boolean { return !this.forbidden.has(zoneId) && !this.operationalClosedZones.has(zoneId); }
  edgeCapacityFactor(edgeId: string): number {
    const edge = this.pack.edges?.[edgeId];
    if (!edge) return 0;
    return Math.min(this.edgeCapacityFactors.get(edgeId) ?? 1, this.zoneCapacityFactors.get(edge.source) ?? 1, this.zoneCapacityFactors.get(edge.destination) ?? 1);
  }

  neighbours(zoneId: string): Array<[string, string]> {
    return this.adjacency.get(zoneId) ?? [];
  }

  edgeCost(edgeId: string, states?: Record<string, ZoneState>, avoid?: Set<string>): [number, number] {
    const edge = this.pack.edges?.[edgeId];
    if (!edge || !this.isEdgeAvailable(edgeId)) return [Number.POSITIVE_INFINITY, Number.POSITIVE_INFINITY];
    const state = states?.[edge.destination] ?? states?.[edge.source];
    const speed = state
      ? Math.max(MIN_SPEED_MS, state.mean_speed_ms)
      : (edge.free_speed_ms?.value ?? FREE_FLOW_SPEED_MS);
    let cost = edge.length_m / speed;
    const travel = cost;
    if (state?.band === 'building') cost *= CONGESTION_WEIGHT;
    else if (state?.band === 'critical') cost *= CRITICAL_WEIGHT;
    if (!isTrustworthy(edge.width_m)) cost *= UNTRUSTED_WIDTH_PENALTY;
    if (avoid?.has(edge.source) || avoid?.has(edge.destination)) cost *= AVOID_PENALTY;
    cost /= this.edgeCapacityFactor(edgeId);
    return [cost, travel];
  }

  route(
    origin: string,
    destination: string,
    states?: Record<string, ZoneState>,
    avoid?: Set<string>,
    prefer?: Set<string>,
    crossingDeadlines?: Record<string, number>,
  ): RouteResult {
    if (states != null || (crossingDeadlines && Object.keys(crossingDeadlines).length > 0)) {
      return this.search(origin, destination, states, avoid, prefer, crossingDeadlines);
    }
    const key = JSON.stringify([origin, destination, [...(avoid ?? [])].sort(), [...(prefer ?? [])].sort()]);
    const cached = this.cache.get(key);
    if (cached) {
      this.cacheHits += 1;
      this.cache.delete(key);
      this.cache.set(key, cached);
      return { ...cached, path: [...cached.path], edge_ids: [...cached.edge_ids] };
    }
    this.cacheMisses += 1;
    const result = this.search(origin, destination, undefined, avoid, prefer, undefined);
    this.cache.set(key, result);
    if (this.cache.size > ASSUMED_ROUTE_CACHE_ENTRIES) {
      const oldest = this.cache.keys().next().value as string | undefined;
      if (oldest != null) this.cache.delete(oldest);
    }
    return { ...result, path: [...result.path], edge_ids: [...result.edge_ids] };
  }

  private search(
    origin: string,
    destination: string,
    states?: Record<string, ZoneState>,
    avoid?: Set<string>,
    prefer?: Set<string>,
    deadlines?: Record<string, number>,
  ): RouteResult {
    const zones = this.pack.zones ?? {};
    if (!(origin in zones))
      return {
        path: [],
        edge_ids: [],
        cost_s: Infinity,
        distance_m: 0,
        eta_s: 0,
        rejected_reason: `unknown origin ${origin}`,
      };
    if (!(destination in zones))
      return {
        path: [],
        edge_ids: [],
        cost_s: Infinity,
        distance_m: 0,
        eta_s: 0,
        rejected_reason: `unknown destination ${destination}`,
      };
    if (!this.isZoneAvailable(destination))
      return {
        path: [],
        edge_ids: [],
        cost_s: Infinity,
        distance_m: 0,
        eta_s: 0,
        rejected_reason: `destination ${destination} is unavailable`,
      };
    if (origin === destination) return { path: [origin], edge_ids: [], cost_s: 0, distance_m: 0, eta_s: 0 };

    const best = new Map([[origin, 0]]);
    const elapsed = new Map([[origin, 0]]);
    const distance = new Map([[origin, 0]]);
    const came = new Map<string, [string, string]>();
    const queue: Array<[number, string]> = [[0, origin]];
    const seen = new Set<string>();

    while (queue.length) {
      queue.sort((a, b) => a[0] - b[0] || a[1].localeCompare(b[1]));
      const [, node] = queue.shift()!;
      if (seen.has(node)) continue;
      seen.add(node);
      if (node === destination) break;
      for (const [next, edgeId] of this.neighbours(node)) {
        if (seen.has(next) || !this.isEdgeAvailable(edgeId) || !this.isZoneAvailable(next)) continue;
        let [cost, travel] = this.edgeCost(edgeId, states, avoid);
        if (prefer?.has(edgeId) || prefer?.has(next)) cost *= PREFER_DISCOUNT;
        const arrival = (elapsed.get(node) ?? 0) + travel;
        if (deadlines?.[edgeId] != null && arrival > deadlines[edgeId]!) continue;
        const candidate = (best.get(node) ?? Infinity) + cost;
        if (candidate < (best.get(next) ?? Infinity)) {
          best.set(next, candidate);
          elapsed.set(next, arrival);
          const edge = this.pack.edges?.[edgeId];
          distance.set(next, (distance.get(node) ?? 0) + (edge?.length_m ?? 0));
          came.set(next, [node, edgeId]);
          queue.push([candidate, next]);
        }
      }
    }

    if (!best.has(destination))
      return {
        path: [],
        edge_ids: [],
        cost_s: Infinity,
        distance_m: 0,
        eta_s: 0,
        rejected_reason: 'no path under current conditions',
      };
    const path = [destination];
    const edgeIds: string[] = [];
    while (path.at(-1) !== origin) {
      const [previous, edgeId] = came.get(path.at(-1)!)!;
      path.push(previous);
      edgeIds.push(edgeId);
    }
    path.reverse();
    edgeIds.reverse();
    return {
      path,
      edge_ids: edgeIds,
      cost_s: best.get(destination)!,
      distance_m: distance.get(destination)!,
      eta_s: elapsed.get(destination)!,
    };
  }

  reachable(origin: string): Set<string> {
    const found = new Set([origin]);
    const stack = [origin];
    while (stack.length) {
      for (const [next] of this.neighbours(stack.pop()!)) {
        if (!found.has(next)) {
          found.add(next);
          stack.push(next);
        }
      }
    }
    return found;
  }
}

File: packages/core/src/crowdflow_core/routing/graph.py
"""Venue graph and routing.

Edge cost is not distance. It is the time the walk actually takes under current
conditions, plus penalties for risk and for edges that are close to their limit:

    cost = travel_time + congestion_penalty + risk_penalty + capacity_penalty

Two things routing must refuse to do, both of which look fine on a static map:

  * route through an edge that is closed, or will have closed by the time the
    walker arrives (D5 — the ETA gate)
  * route over geometry it does not trust (D6 — provenance)

Static routes are cached. See `route` for why that is safe and where it stops.
"""

from __future__ import annotations

import heapq
import math
from dataclasses import dataclass, field, replace

from crowdflow_contracts import CircuitPack, FREE_FLOW_SPEED_MS, LOSBand, ZoneState

from ..state.flow import MIN_SPEED_MS

CONGESTION_WEIGHT = 2.5
"""Multiplier on travel time when an edge is in the BUILDING band. Routing should
prefer a longer clear path over a shorter slow one, but not absurdly so."""

CRITICAL_WEIGHT = 12.0
"""BUILDING is discouraged; CRITICAL is nearly forbidden. Not infinite, because a
graph with no path at all is worse than a bad path."""

UNTRUSTED_WIDTH_PENALTY = 1.15
"""Small tax on edges whose width is assumed. Their band is provisional, so their
cost estimate is too; prefer corroborated geometry where the choice is close."""

AVOID_PENALTY = 25.0
"""Multiplier on an edge touching a zone the operator asked walkers to avoid.
ASSUMED. Large enough that any plausible detour wins, finite because an advisory
is not a closure: if avoiding a zone would strand someone, they still get a path
through it. Compare CRITICAL_WEIGHT, which is the same argument for a measured
condition rather than an instruction."""

PREFER_DISCOUNT = 0.6
"""Multiplier on an edge the operator asked walkers to prefer. ASSUMED. A
discount rather than a rewrite: preferring a route must be able to lose to a
much shorter alternative, or an advisory becomes a diversion nobody chose."""


@dataclass
class RouteResult:
    path: list[str] = field(default_factory=list)
    cost_s: float = math.inf
    distance_m: float = 0.0
    eta_s: float = 0.0
    rejected_reason: str | None = None

    @property
    def found(self) -> bool:
        return bool(self.path) and self.rejected_reason is None


class VenueGraph:
    """Adjacency over a CircuitPack, with dynamic per-tick costs.

    Rebuilt when the session state changes (crossings open and close); re-costed
    every tick (density moves). Those are different frequencies, so they are
    different operations.
    """

    def __init__(self, pack: CircuitPack, session_state: str | None = None) -> None:
        self.pack = pack
        self.session_state = session_state
        self._adj: dict[str, list[tuple[str, str]]] = {}
        self._closed: set[str] = set()
        self._forbidden: set[str] = set()
        self._route_cache: dict[
            tuple[str, str, frozenset[str], frozenset[str]], RouteResult
        ] = {}
        self.cache_hits = 0
        self.cache_misses = 0
        self.rebuild(session_state)

    # -- structure ---------------------------------------------------------

    def rebuild(self, session_state: str | None) -> None:
        """Recompute which edges exist at all, given the session state (D5).

        Drops the route cache. A cached path computed while a crossing was open
        is not merely stale after it shuts — it routes people at a closed
        crossing, which is the exact failure the ETA gate exists to prevent.
        Invalidation lives here, in the one method that can change the edge set,
        so it cannot be forgotten at a call site.
        """
        self.session_state = session_state
        self._route_cache.clear()
        self._closed = set()
        for crossing in self.pack.crossings.values():
            if not crossing.availability.is_open_during(session_state):
                self._closed.add(crossing.edge_id)

        # Forbidden zones are REMOVED, not penalised. `avoid` is a preference
        # expressed as a cost multiplier, which is right for "steer traffic away"
        # and wrong for "never route through a live-circuit working position": a
        # multiplier still yields a path when it is the only one, and a path is
        # what the caller then acts on. If the only way runs through a marshal
        # post, the honest answer is that there is no way.
        self._forbidden = set(self.pack.constraints.never_route_through)

        self._adj = {zid: [] for zid in self.pack.zones}
        for eid, e in self.pack.edges.items():
            if eid in self._closed:
                continue
            if e.source in self._forbidden or e.destination in self._forbidden:
                continue
            self._adj.setdefault(e.source, []).append((e.destination, eid))
            if e.bidirectional:
                self._adj.setdefault(e.destination, []).append((e.source, eid))

    @property
    def closed_edges(self) -> set[str]:
        return set(self._closed)

    @property
    def forbidden_zones(self) -> set[str]:
        """Zones no route may traverse. Structurally absent from the graph."""
        return set(self._forbidden)

    def path_violations(self, path: list[str]) -> list[str]:
        """Forbidden zones a path traverses. Should always be empty — this is the
        independent check that the structural exclusion actually held."""
        return [z for z in path if z in self._forbidden]

    def neighbours(self, zone_id: str) -> list[tuple[str, str]]:
        return self._adj.get(zone_id, [])

    # -- cost --------------------------------------------------------------

    def edge_cost(
        self,
        edge_id: str,
        states: dict[str, ZoneState] | None = None,
        avoid: set[str] | None = None,
    ) -> tuple[float, float]:
        """Return (cost_seconds, travel_time_seconds) for traversing an edge."""
        e = self.pack.edges[edge_id]
        state = (states or {}).get(e.destination) or (states or {}).get(e.source)

        speed = MIN_SPEED_MS
        band = LOSBand.NOMINAL
        if state is not None:
            speed = max(MIN_SPEED_MS, state.mean_speed_ms)
            band = state.band
        else:
            speed = e.free_speed_ms.value if e.free_speed_ms else FREE_FLOW_SPEED_MS

        travel = e.length_m / speed
        cost = travel
        if band is LOSBand.BUILDING:
            cost *= CONGESTION_WEIGHT
        elif band is LOSBand.CRITICAL:
            cost *= CRITICAL_WEIGHT
        if not e.width_m.is_trustworthy:
            cost *= UNTRUSTED_WIDTH_PENALTY
        if avoid and (e.source in avoid or e.destination in avoid):
            cost *= AVOID_PENALTY
        return cost, travel

    # -- search ------------------------------------------------------------

    def _heuristic(self, a: str, b: str) -> float:
        za, zb = self.pack.zones.get(a), self.pack.zones.get(b)
        if not za or not zb:
            return 0.0
        d = math.dist((za.position.x, za.position.y), (zb.position.x, zb.position.y))
        # Optimistic: free-flow walk on the straight line, so A* stays admissible.
        return d / FREE_FLOW_SPEED_MS

    def route(
        self,
        origin: str,
        destination: str,
        states: dict[str, ZoneState] | None = None,
        avoid: set[str] | None = None,
        prefer: set[str] | None = None,
        depart_at: float = 0.0,
        crossing_deadlines: dict[str, float] | None = None,
    ) -> RouteResult:
        """A* over dynamic cost, with the ETA gate on time-limited edges.

        crossing_deadlines maps edge_id -> seconds from depart_at after which
        that edge closes. A path is rejected if the walker would arrive at an
        edge after it shuts: routing someone toward a crossing that closes before
        they get there manufactures the queue it was trying to prevent.

        **Memoised, but only where the answer is a pure function of the graph.**
        Six thousand simulated agents share on the order of tens of distinct
        (origin, destination, avoid, prefer) requests, and the intervention
        engine forks the whole world once per candidate — so the same handful of
        searches over a 1,875-node graph was being repeated tens of thousands of
        times per decision. Forks share this graph, so they share the cache.

        Two boundaries keep the memo honest, and both are load-bearing:

          * `states` is the per-tick density field. It changes every tick, so a
            cached cost would be a cost from a crowd that has since moved. A
            call that passes states is never cached and never served from cache.
          * `crossing_deadlines` makes the answer depend on when the walker
            leaves, which is not in the key. Same treatment.

        Invalidation on structural change lives in `rebuild`.
        """
        if states is not None or crossing_deadlines:
            return self._search(
                origin, destination, states, avoid, prefer, crossing_deadlines
            )

        key = (origin, destination, frozenset(avoid or ()), frozenset(prefer or ()))
        cached = self._route_cache.get(key)
        if cached is None:
            self.cache_misses += 1
            cached = self._search(origin, destination, None, avoid, prefer, None)
            self._route_cache[key] = cached
        else:
            self.cache_hits += 1
        # Hand back a copy: callers own their path list, and one that mutated it
        # would corrupt every later hit rather than fail visibly.
        return replace(cached, path=list(cached.path))

    @property
    def route_cache_size(self) -> int:
        return len(self._route_cache)

    def _search(
        self,
        origin: str,
        destination: str,
        states: dict[str, ZoneState] | None,
        avoid: set[str] | None,
        prefer: set[str] | None,
        crossing_deadlines: dict[str, float] | None,
    ) -> RouteResult:
        """The A* itself. Pure in its arguments — which is what makes `route`
        cacheable, and why the cache wrapper is separate from the search."""
        if origin not in self.pack.zones:
            return RouteResult(rejected_reason=f"unknown origin {origin!r}")
        if destination not in self.pack.zones:
            return RouteResult(rejected_reason=f"unknown destination {destination!r}")
        if origin == destination:
            return RouteResult(path=[origin], cost_s=0.0)

        deadlines = crossing_deadlines or {}
        prefer = prefer or set()

        best: dict[str, float] = {origin: 0.0}
        elapsed: dict[str, float] = {origin: 0.0}
        dist: dict[str, float] = {origin: 0.0}
        came: dict[str, tuple[str, str]] = {}
        heap: list[tuple[float, str]] = [(self._heuristic(origin, destination), origin)]
        seen: set[str] = set()

        while heap:
            _, node = heapq.heappop(heap)
            if node in seen:
                continue
            seen.add(node)
            if node == destination:
                break

            for nxt, eid in self.neighbours(node):
                if nxt in seen:
                    continue
                cost, travel = self.edge_cost(eid, states, avoid)
                if eid in prefer or nxt in prefer:
                    cost *= PREFER_DISCOUNT

                arrive = elapsed[node] + travel
                if eid in deadlines and arrive > deadlines[eid]:
                    continue  # would arrive after it closes — the ETA gate

                g = best[node] + cost
                if g < best.get(nxt, math.inf):
                    best[nxt] = g
                    elapsed[nxt] = arrive
                    dist[nxt] = dist[node] + self.pack.edges[eid].length_m
                    came[nxt] = (node, eid)
                    heapq.heappush(heap, (g + self._heuristic(nxt, destination), nxt))

        if destination not in best:
            return RouteResult(rejected_reason="no path under current conditions")

        path: list[str] = [destination]
        while path[-1] != origin:
            path.append(came[path[-1]][0])
        path.reverse()

        return RouteResult(
            path=path,
            cost_s=best[destination],
            distance_m=dist[destination],
            eta_s=elapsed[destination],
        )

    def reachable(self, origin: str) -> set[str]:
        """Connectivity check — used by validation and by safety."""
        seen = {origin}
        stack = [origin]
        while stack:
            for nxt, _ in self.neighbours(stack.pop()):
                if nxt not in seen:
                    seen.add(nxt)
                    stack.append(nxt)
        return seen
